ivy/stateful/module.py

```diff
@@ -362,6 +364,7 @@ def register_parameter(self, name, value):
             Value of the buffer
         """
         self._v.update({name: value})
+        super().__setattr__(name, value)
```

Comment:
I had intentionally avoided doing a `__setattr__` here actually so that we don't end up with a module and a parameter having the same name, leading to confusion. So for e.g. if you set something as a module `model.mod1 = <some ivy module>` and then do a `register_parameter("mod1", <array>)` then `model.mod1` becomes the parameter rather than the module itself 😅
I know that `torch` does it that way but with `torch` they have an explicit [`add_module`](https://github.com/pytorch/pytorch/blob/8933ff35953908545cb410d736a6c4a3a7761c54/torch/nn/modules/module.py#L596) method, but adding it and all the other validation for checking if the same name exists in the module, etc. might end up taking more time than adding somewhat a restriction where `model.mod1` will always be a module and `model.v.mod1` will always be a variable if that makes sense? Same applies to buffers as well

Reply:
> I had intentionally avoided doing a `__setattr__` here actually so that we don't end up with a module and a parameter having the same name, leading to confusion. So for e.g. if you set something as a module `model.mod1 = <some ivy module>` and then do a `register_parameter("mod1", <array>)` then `model.mod1` becomes the parameter rather than the module itself 😅 I know that `torch` does it that way but with `torch` they have an explicit [`add_module`](https://github.com/pytorch/pytorch/blob/8933ff35953908545cb410d736a6c4a3a7761c54/torch/nn/modules/module.py#L596) method, but adding it and all the other validation for checking if the same name exists in the module, etc. might end up taking more time than adding somewhat a restriction where `model.mod1` will always be a module and `model.v.mod1` will always be a variable if that makes sense? Same applies to buffers as well

There are two ways we can deal with this then. One was this approach which I agree might not be the most robust one to go for. We can however do what `torch` [does](https://github.com/pytorch/pytorch/blob/8933ff35953908545cb410d736a6c4a3a7761c54/torch/nn/modules/module.py#L1676-L1687) by modifying our `__getattribute__` method such that whenever someone registers a parameter like `register_parameter("bias", ....)` and then subsequently fetches that using `self.bias`, we should still be able to retrieve that paramter from the module object. `torch` does this explicitly for parameters, buffers and modules. It is pertinent that if a parameter was registered i.e. added to `._v`, we still should be able to retrieve it by simply doing `model.bias` rather than `model.v.bias` if we want to superset `torch`'s behaviour here. What do you think? @vedpatwardhan

Reply:
Same is the [case](https://github.com/pytorch/pytorch/blob/8933ff35953908545cb410d736a6c4a3a7761c54/torch/nn/modules/module.py#L1750-L1758) with `delattr` actually

Reply:
Yeah exactly, the main difference between our design and theirs is that with `torch`,
```
import torch
from torch import nn


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.mod1 = nn.Linear(10, 10)
        self.mod2 = nn.Linear(10, 10)
    
    def forward(self, x):
        return self.mod2(self.mod1(x))


model = Model()
print(model._parameters)
print(model._modules)
print(list(model.parameters()))
```
you'd see that `mod1` isn't a part of the `_parameters` but is a part of the `_modules`, but with our design with an equivalent `ivy.Module`, `mod1` would be a part of both the `_v` and the `_module_dict`, so we can't really distinguish between them when doing a `__getattr__` or a `__setattr__` or a `__delattr__`. 
So in order to replicate their design we'd need a `Parameter` class for e.g. to distinguish between a regular array and a parameter, build modules and find variables from the `_module_dict` everytime we try to access the parameter dict (if our `self._v` is also empty like theirs) and so on which is why I think the current minimal structure is better for now. 
Having said that, I do think we can still implement something like that in the frontends because we do mimic their design there 😄

Reply:
@vedpatwardhan, that's the reason why I came to you asking the other day if there's a way we can distinguish between regular arrays and model parameters in `ivy` 😄

Reply:
Yeah so in `ivy` you could distinguish them based on whether or not they are a part of the `model._v`, if an array's not a part of the `model._v` then it's a regular array 😄
So basically whenever a `torch.nn.Parameter` is created in the frontend, the corresponding `ivy.Array` would be a part of the `model._v` of the `ivy.Module` if that makes sense?

Reply:
@vedpatwardhan Also, all this was implemented in the frontends [previously](https://github.com/unifyai/ivy/blob/d9e8e49e16965a7c0dff22c0cfaad4f4422f6013/ivy/functional/frontends/torch/nn/modules/module.py#L717-L799) but it doesn't make sense to keep such lengthy implementations in the frontends if we want to eventually design `ivy.Module` to superset `torch.nn.Module` and other stateful classes anyways. I'm not sure what's the best way to go about this in the interest of time. Also that the work I'm doing with the transformations relies heavily on the changes in this PR

Reply:
Actually we are already supersetting the stuff being added to the `torch` frontend module in the sense that the `._v` is present and the `_module_dict` is present in the `ivy.Module` and we can add a new module or a new parameter, etc. The main difference is the way we use it. I do get the point that it might be a bit cluttered in the frontends, but unfortunately if the `torch.nn.Module` itself has those condition checks and so on, which creates the need to have those in the frontend 😅

Reply:
@vedpatwardhan Thing I'm trying to understand is it makes perfect sense if someone does `self.register_paramter("bias", ..) and then subsequently use `self.bias` simply afterwards even if it is just `ivy.Module` and not `torch.nn.Module`. Wouldn't you think?

Reply:
> Actually I didn't understand. Why would `mod1` be a part of both `_v` and `_module_dict` here? I've reverted the `__setattr__` method entirely so not sure why this would still happen? We are not calling `register_parameter` here. This is a simply `__setattr__` call, no? @vedpatwardhan

Probably I'm missing something but the `__setattr__` call seems to be there on [this](https://github.com/unifyai/ivy/pull/27403/files#diff-9d3cd8360ac6368fb7cfebdd75ef04230b2963749af122e8515e44ea6ddf3971R367) line which would set the name of the variable as an attribute of the object ig 😅

Reply:
> @vedpatwardhan Thing I'm trying to understand is it makes perfect sense if someone does `self.register_paramter("bias", ..)` and then subsequently use `self.bias` simply afterwards even if it is just `ivy.Module` and not `torch.nn.Module`. Wouldn't you think?

Yeah I do think that would be great (and more intuitive) but the problem is we need to do a fairly large refactor into the `Module` to get something close to it working, like changing our design such that `self._v` only contains the parameters that are registered using `register_parameter` and not the ones in the modules, then we'd need to incorporate this into the `model.v` property so that we `_find_variables` everytime the property is called, and add all the checks to avoid duplicates, etc. Basically we'd need to get our `ivy.Module` restructured quite a bit so that closely aligns with `torch.nn.Module` 😅